### PR TITLE
fix typing in StacApiIO

### PR DESCRIPTION
**Pull request by philvarner**
**Related Issue(s):** #189 

**Description:**

- fix typing in StacApiIO class

**PR Checklist:**

- [X] Code is formatted
- [X] Tests pass
- [X] Changes are added to the [CHANGELOG](https://github.com/stac-utils/pystac-api-client/blob/main/CHANGELOG.md)

### CODE REVIEW

**Review by matthewhanson — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ### Added
 
+- Significantly improved type hints
```
Comment (gadomski) [inline CHANGELOG.md:11]: Needs changelog link

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -25,15 +25,15 @@ repos:
...
-#          - types-requests
-#          - types-python-dateutil
+  - repo: https://github.com/pre-commit/mirrors-mypy
+    rev: v0.960
+    hooks:
+      - id: mypy
+        files: ".*\\.py$"
+        exclude: ^tests/.*$
+        args:
+          - --ignore-missing-imports
```
Comment (gadomski) [inline .pre-commit-config.yaml:35]: I _think_ we can remove this.

File: pystac_client/stac_api_io.py [hunk 3]
```diff
@@ -50,33 +50,23 @@ def __init__(
         # TODO - this should super() to parent class
         self.session = Session()
         self.session.headers.update(headers or {})
-        self.session.params.update(parameters or {})
+
+        if hasattr(self.session.params, "update"):
```
Comment (gadomski) [inline pystac_client/stac_api_io.py:54]: How could the `Session` not have a `dict` as its `params` attribute? Is it a `requests` version thing?

File: pystac_client/stac_api_io.py [hunk 4]
```diff
@@ -50,33 +50,23 @@ def __init__(
...
-        if isinstance(source, str):
-            href = source
-            if bool(urlparse(href).scheme):
-                return self.request(href, *args, parameters=parameters, **kwargs)
-            else:
-                with open(href) as f:
-                    href_contents = f.read()
-                return href_contents
-        elif isinstance(source, Link):
+        if isinstance(source, Link):
```
Comment (gadomski) [inline pystac_client/stac_api_io.py:69]: Per the typing, `pystac.link.HREF` is either a `str` or a `PathLike`, so can't be a `Link`?

**Review by gadomski — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (philvarner) [reply CHANGELOG.md:11]: I left it out since there are 6-8 PRs

File: pystac_client/stac_api_io.py (hunk 3 above)
Comment (philvarner) [reply pystac_client/stac_api_io.py:54]: yeah, params is typed as the completely incomprehensible union of several nested layers of type hints. The other option is to ignore this, which is probleably more reasonable to do

File: pystac_client/stac_api_io.py (hunk 4 above)
Comment (philvarner) [reply pystac_client/stac_api_io.py:65]: `class Link(PathLike):`

Comment (gadomski) [reply pystac_client/stac_api_io.py:65]: Whoa, did not realize that. Don't know if I like it, but not our problem here. Disregard, thanks.

**Review by gadomski — APPROVED** [review]

**Pull request merged**